### PR TITLE
[GEOS-9522] SldService generate overlapping rules

### PR DESCRIPTION
**Pull request by taba90**
<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Committs changing the REST API, or any configuration object, should check it the REST API docs (Swagger YAML files and classic documentation) need to be updated.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by taba90** [comment]
Related JIRA ticket: https://osgeo-org.atlassian.net/browse/GEOS-9522

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java [hunk 1]
```diff
@@ -1987,4 +1992,134 @@ private void checkRuleLineSymbolizer(Rule rule, String color) {
...
+                        + "attribute=bar&ramp=red&method=quantile&intervals=2&open=true";
+        Document dom = getAsDOM(restPath, 200);
+        print(dom);
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        print(dom, baos);
+        String resultXml = baos.toString().replace("\r", "").replace("\n", "");
+        Rule[] rules =
+                checkSLD(resultXml.replace("<Rules>", sldPrefix).replace("</Rules>", sldPostfix));
+        assertTrue(rules.length == 2);
+        assertTrue(rules[0].getFilter() instanceof PropertyIsEqualTo);
```
Comment (aaime) [inline src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: I'd also check that the second rule here, and that they won't overlap?

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java [hunk 2]
```diff
@@ -1987,4 +1992,134 @@ private void checkRuleLineSymbolizer(Rule rule, String color) {
...
+        ByteArrayOutputStream baosJenks = new ByteArrayOutputStream();
+        print(domJenks, baosJenks);
+        String resultXmlJenks = baosJenks.toString().replace("\r", "").replace("\n", "");
+        Rule[] rulesJenks =
+                checkSLD(
+                        resultXmlJenks
+                                .replace("<Rules>", sldPrefix)
+                                .replace("</Rules>", sldPostfix));
+        assertTrue(rulesJenks.length == 2);
+        assertTrue(rulesJenks[0].getFilter() instanceof PropertyIsEqualTo);
```
Comment (aaime) [inline src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: Same as above.

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java [hunk 3]
```diff
@@ -1987,4 +1992,134 @@ private void checkRuleLineSymbolizer(Rule rule, String color) {
...
+                        resultXmlJenks
+                                .replace("<Rules>", sldPrefix)
+                                .replace("</Rules>", sldPostfix));
+        assertTrue(rulesJenks.length == 2);
+        Rule firstJenks = rulesJenks[0];
+        Rule secondJenks = rulesJenks[1];
+        checkNotOverlappingRules(firstJenks, secondJenks);
+    }
+
+    private void checkNotOverlappingRules(Rule first, Rule second) {
```
Comment (aaime) [inline src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: Name suggests a general non overlap test, but in fact it's pretty specific and does not test what happens in the first rule filter.
Suggestion for a different approach:

```
SimpleFeatureType ft = (SimpleFeatureType)= getCatalog().getFeatureTypeByName("ClassificationPoints2").getFeatureType();
SimpleFeature feature = DataUtilities.create(ft, "=1|2.0|POINT(4 2.5)");
assertTrue(first.getFilter().evaluate(feature));
assertFalse(second.getFilter().evaluate(feature));
```

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java [hunk 4]
```diff
@@ -470,4 +476,29 @@ private Expression normalizeProperty(
...
+     * @return
+     */
+    private Filter getNotOverlappingFilter(
+            int currentIdx, RangedClassifier groups, Expression att, List<Rule> list) {
+        Filter f;
+        if (currentIdx > 0) {
+            Rule prevRule = list.get(list.size() - 1);
+            Object currMin = groups.getMin(currentIdx);
+            Object prevMin = groups.getMin(currentIdx - 1);
+            if (!prevMin.equals(currMin) || !(prevRule.getFilter() instanceof PropertyIsEqualTo))
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: Isn't this redundant? If the prev min is the same as the current min, the previous filter must include the value... or not? Asking cause I'm not sure, but if it does, you could reduce the list of arguments of this method.

Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: Indeed it is, removed the check as well as the param.

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java (hunk 3 above)
Comment (taba90) [reply src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: Thanks for the suggestion, applied.

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java (hunk 2 above)
Comment (taba90) [reply src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: fixed

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java (hunk 1 above)
Comment (taba90) [reply src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: fixed

**Pull request merged**